File: 2025/07/day7.py
#!/usr/bin/env python3
from functools import cache
start = None
beams = set()
split = 0
with open("input.txt") as f:
    for line in f:
        if not start:
            start = line.index("S")
            beams.add(start)
        else:
            ss = [ i for i, x in enumerate(line) if x == "^" ]
            for splitter in ss:
                if splitter in beams:
                    split += 1
                    beams.remove(splitter)
                    beams.add(splitter - 1)
                    beams.add(splitter + 1)
print(split)

grid = []
with open("input.txt") as f:
    grid = [ [ c == "^" for c in line ] for line in f ]

@cache
def count_the_ways(y, x):
    depth = y + 1
    while depth < len(grid):
        if grid[depth][x]:
            return count_the_ways(depth, x - 1) + count_the_ways(depth, x + 1)
        depth += 1
    return 1

print(count_the_ways(0, start))

File: 2025/07/day7.swift
#!/usr/bin/env swift
import Foundation

let inputFile = URL(filePath: "input.txt")

let data = try! Data(contentsOf: inputFile)
let lines = data.split(separator: UInt8(ascii: "\n"))

let splitters = lines.dropFirst().map { line in
    line.map { $0 == UInt8(ascii: "^") }
}

var split = 0
var beamCounts = lines.first!.map { $0 == UInt8(ascii: "S") ? 1 : 0 }

for row in splitters {
    for i in row.indices where row[i] && beamCounts[i] > 0 {
        split += 1
        beamCounts[i - 1] += beamCounts[i]
        beamCounts[i + 1] += beamCounts[i]
        beamCounts[i] = 0
    }
}

print(split)
print(beamCounts.reduce(0, +))
